_episodes/l1-01-introduction.md

```diff
@@ -1,18 +1,222 @@
+an impressive looking collection of tests that have very little value at all.
+What should be the aim therefore when developing software tests?
+
+In practice this is difficult to define universally but one useful mantra is
+that good tests ***thoroughly exercise critical code***. One way to achieve this
+is to design test examples of increasing complexity that cover the most general
+case the unit should encounter. Also try to consider examples of special or edge
+cases that your function needs to handle especially.
+
+A useful quantitative metric to consider is **test coverage**. Using additional
+tools it is possible to determine, on a line-by-line basis, the proportion of a
+codebase that is being exercised by its tests. This can be useful to ensure, for
+instance, that all logical branching points within the code are being used by
+the test inputs.
+
+> ## Testing and Coverage
+>
+> Consider the following C++ function:
+>
+> ```cpp
+> /*Returns the n'th term of the Fibonacci sequence.*/
+> int recursive_fibonacci(int n)
+> {
+>     if (n <= 1) {
+>         return n;
+>     } else {
+>         return recursive_fibonacci(n - 1) + recursive_fibonacci(n - 2);
+>     }
+> }
+> ```
+>
+> Try to think up some test cases of increasing complexity, there are four
+> distinct cases worth considering. What input value would you use for each case
+> and what output value would you expect? Which lines of code will be exercised
+> by each test case? How many cases would be required to reach 100% coverage?
+>
+> For convenience, some initial terms from the Fibonacci sequence are given
+> below:
+> 0, 1, 1, 2, 3, 5, 8, 13, 21
+>
+> > ## Solution
+> >
+> > ### Case 1 - Use either 0 or 1 as input
+> >
+> > **Correct output:*- Same as input
+> > **Coverage:*- First section of if-block
+> > **Reason:*- This represents the simplest possible test for the function. The
+> > value of this test is that it exercises only the special case tested for by
+> > the if-block.
+> >
+> > ### Case 2 - Use a value > 1 as input
+> >
+> > **Correct output:*- Appropriate value from the Fibonacci sequence
+> > **Coverage:*- All of the code
+> > **Reason:*- This is a more fully fledged case that is representative of the
+> > majority of the possible range of input values for the function. It covers
+> > not only the special case represented by the first if-block but the general
+> > case where recursion is invoked.
+> >
+> > ### Case 3 - Use a negative value as input
+> >
+> > **Correct output:*- Depends...
+> > **Coverage:*- First section of if-block
+> > **Reason:*- This represents the case of a possible input to the function
+> > that is outside of its intended usage. At the moment the function will just
+> > return the input value, but whether this is the correct behaviour depends on
+> > the wider context in which it will be used. It might be better for this type
+> > of input value to cause an error to be raised however. The value of this
```

Comment:
```suggestion
> > the wider context in which it will be used. It might be better for this type
> > of input value to throw an exception, however. The value of this
```

This tends to be the terminology used for C++.